rspec-tools/rspec_tools/validation/description.py

```diff
@@ -1,55 +1,160 @@
+ACCEPTED_FRAMEWORK_NAMES: Final[list[str]] = parse_names('docs/header_names/allowed_framework_names.adoc')
+
+[
+  SECTIONS,
+  OPTIONAL_SECTIONS
+  ] = parse_education_section_names('docs/header_names/education_format_names.adoc')
+
+
+
+def intersection(lst1, lst2):
+  lst3 = [value for value in lst1 if value in lst2]
+  return lst3
+def difference(lst1, lst2):
+  return list(set(lst1) - set(lst2))
 
 def validate_section_names(rule_language: LanguageSpecificRule):
+  """Validates all h2-level section names"""
+
   descr = rule_language.description
-  for h2 in descr.find_all('h2'):
-    name = h2.text.strip()
-    if name not in ACCEPTED_SECTION_NAMES:
-      raise RuleValidationError(f'Rule {rule_language.id} has unconventional header "{name}"')
+  h2_titles = list(map(lambda x: x.text.strip(), descr.find_all('h2')))
+
+  education_titles = intersection(h2_titles, list(SECTIONS.keys()) + list(OPTIONAL_SECTIONS.keys()))
+  if education_titles:
+    # we're using the education format
+    validate_how_to_fix_it_sections_names(rule_language, h2_titles)
+    missing_titles = difference(list(SECTIONS.keys()), education_titles)
+    # we handled "how to fix it" sections above
+    missing_titles_without_how_to_fix_its = [ s for s in missing_titles if not HOW_TO_FIX_IT_REGEX.match(s)]
+    if missing_titles_without_how_to_fix_its:
+      # when using the progressive education format, we need to have all its mandatory titles
+      raise RuleValidationError(f'Rule {rule_language.id} is missing the "{missing_titles_without_how_to_fix_its[0]}" section')
+  else:
+    # we're using the legacy format
+    for title in h2_titles:
+      if title not in LEGACY_SECTION_NAMES:
+        raise RuleValidationError(f'Rule {rule_language.id} has an unconventional header "{title}"')
+
+def validate_how_to_fix_it_sections_names(rule_language: LanguageSpecificRule, h2_titles: list[str]):
+  how_to_fix_it_sections = [ s for s in h2_titles if HOW_TO_FIX_IT_REGEX.match(s) ]
+  if len(how_to_fix_it_sections) > 6:
+    raise RuleValidationError(f'Rule {rule_language.id} has more than 6 "{HOW_TO_FIX_IT}" sections. Please ensure this limit can be increased with PM/UX teams')
+  if not how_to_fix_it_sections:
+    raise RuleValidationError(f'Rule {rule_language.id} is missing a "{HOW_TO_FIX_IT}" section')
+  if HOW_TO_FIX_IT in how_to_fix_it_sections and len(how_to_fix_it_sections) > 1:
+    raise RuleValidationError(f'Rule {rule_language.id} is mixing "{HOW_TO_FIX_IT}" with "How to fix it in FRAMEWORK NAME" sections. Either use a single "{HOW_TO_FIX_IT}" or one or more "How to fix it in FRAMEWORK"')
+  framework_sections_seen = set()
+  for section_name in how_to_fix_it_sections:
+    validate_how_to_fix_it_framework(section_name, rule_language, framework_sections_seen)
+
+def validate_how_to_fix_it_framework(section_name, rule_language, framework_sections_seen):
```

Comment:
```suggestion
  for section_name in how_to_fix_it_sections:
    validate_how_to_fix_it_framework(section_name, rule_language)

def validate_how_to_fix_it_framework(section_name, rule_language):
    framework_sections_seen = set()
```
Why do you keep a single set for all "How to fix it " sections?

Reply:
`validate_how_to_fix_it_framework` checks a single `How to fix it in...` section, so I must collect the framework names one level above.

Reply:
I see now. I find this pretty hard to follow. Moreover, you are essentially passing validation of the "how to fix it" section name to the `validate_how_to_fix_it_framework` callee.
Since you already have the list of all "how to fix it in" sections: `how_to_fix_it_sections`, why not check duplication of section names in `validate_how_to_fix_it_sections_names`? For example:
```
  duplicate_names = [x for x in how_to_fix_it_sections if how_to_fix_it_sections.count(x) > 1] # O(n*n) is fine, given n <= 6
  if 0 < duplicate_names.len():
    raise RuleValidationError(f'Rule {rule_language.id} has duplicate "{HOW_TO_FIX_IT}" sections "{duplicate_names}"')
```

Reply:
done